File: backend/app/cc98.py
from __future__ import annotations

import asyncio
from dataclasses import dataclass
from typing import Any, Protocol

import httpx

from .config import settings
from .models import Reply, TopicSummary


class CC98Error(Exception):
    def __init__(self, message: str, *, kind: str = "network", status_code: int | None = None) -> None:
        super().__init__(message)
        self.kind = kind
        self.status_code = status_code


class CC98Client(Protocol):
    async def validate_token(self, token: str) -> None: ...
    async def search_topics(self, token: str, query: str) -> list[TopicSummary]: ...
    async def get_replies(self, token: str, topic_id: str, *, page: int = 1, page_size: int = 100) -> tuple[list[Reply], bool]: ...


@dataclass
class HttpCC98Client:
    api_base: str = settings.cc98_api_base
    timeout: float = settings.cc98_timeout_seconds

    async def _request(self, token: str, method: str, path: str, **kwargs: Any) -> Any:
        headers = {"Authorization": f"Bearer {token}", "Accept": "application/json"}
        try:
            async with httpx.AsyncClient(
                base_url=self.api_base,
                timeout=self.timeout,
                trust_env=settings.cc98_trust_env,
            ) as client:
                response = await client.request(method, path, headers=headers, **kwargs)
        except httpx.HTTPError as exc:
            raise CC98Error("CC98 网络请求失败", kind="network") from exc
        if response.status_code in (401, 403):
            raise CC98Error("CC98 会话无效或无权限", kind="auth", status_code=response.status_code)
        if response.status_code == 429:
            raise CC98Error("CC98 请求过于频繁", kind="rate_limit", status_code=429)
        if response.status_code >= 400:
            raise CC98Error("CC98 服务返回错误", kind="service", status_code=response.status_code)
        try:
            return response.json()
        except ValueError as exc:
            raise CC98Error("CC98 返回了无法解析的数据", kind="service") from exc

    async def validate_token(self, token: str) -> None:
        await self._request(token, "GET", "/user/profile")

    async def search_topics(self, token: str, query: str) -> list[TopicSummary]:
        data = await self._request(token, "GET", settings.cc98_topic_search_path, params={"keyword": query})
        items = data.get("data", data) if isinstance(data, dict) else data
        if not isinstance(items, list):
            items = items.get("items", []) if isinstance(items, dict) else []
        return [_topic_from_json(item) for item in items if isinstance(item, dict)]

    async def get_replies(self, token: str, topic_id: str, *, page: int = 1, page_size: int = 100) -> tuple[list[Reply], bool]:
        data = await self._request(token, "GET", f"/topic/{topic_id}/post", params={"page": page, "pageSize": page_size})
        raw = data.get("data", data) if isinstance(data, dict) else data
        if isinstance(raw, dict):
            items = raw.get("items", raw.get("posts", []))
            has_more = bool(raw.get("hasMore", raw.get("has_more", len(items) >= page_size)))
        else:
            items, has_more = raw if isinstance(raw, list) else [], False
        return [_reply_from_json(item) for item in items if isinstance(item, dict)], has_more


def _topic_from_json(item: dict[str, Any]) -> TopicSummary:
    topic_id = str(item.get("id", item.get("topicId", item.get("topic_id", ""))))
    url = item.get("url") or item.get("link") or f"https://www.cc98.org/topic/{topic_id}"
    return TopicSummary(id=topic_id, title=str(item.get("title", item.get("subject", ""))), board=item.get("boardName", item.get("board")), published_at=item.get("postTime", item.get("publishedAt", item.get("createTime"))), reply_count=item.get("replyCount", item.get("replies")), url=str(url))


def _reply_from_json(item: dict[str, Any]) -> Reply:
    return Reply(id=str(item.get("id", item.get("postId", ""))), author=item.get("authorName", item.get("author")), content=str(item.get("content", item.get("text", ""))), floor=item.get("floor", item.get("index")), published_at=item.get("postTime", item.get("publishedAt")))


async def fetch_all_replies(client: CC98Client, token: str, topic_id: str, *, max_pages: int = 20) -> list[Reply]:
    all_replies: list[Reply] = []
    page = 1
    while True:
        replies, has_more = await client.get_replies(token, topic_id, page=page)
        all_replies.extend(replies)
        if not has_more or not replies or page >= max_pages:
            return all_replies
        page += 1
        await asyncio.sleep(0)

File: frontend/src/extension/cc98.ts
import {
  SourceError,
  type Candidate,
  type PageContext,
  type RatePolicy,
  type SearchHit,
  type SearchPage,
  type SearchSourceAdapter,
  type SearchSourceSession,
  type SourceCapabilities,
} from './types';
import { normalizeText } from './planner';

const PAGE_SIZE = 20;

export const CC98_RATE_POLICY: RatePolicy = {
  maxSearchCalls: 30,
  minRequestIntervalMs: 2000,
};

export const CC98_CAPABILITIES: SourceCapabilities = {
  searchSurface: 'title',
  querySyntax: 'plain-keyword',
};

type FetchLike = (input: string, init?: RequestInit) => Promise<Response>;

export function readCc98AccessToken(storage: Pick<Storage, 'getItem'>): string | null {
  const expiration = storage.getItem('accessToken_expirationTime');
  if (expiration && Date.now() >= Number.parseInt(expiration, 10) * 1000) return null;
  const stored = storage.getItem('accessToken');
  if (!stored) return null;
  const token = stored.startsWith('str-') ? stored.slice(4) : stored;
  return /^Bearer\s+\S+$/iu.test(token) ? token : null;
}

function topicList(payload: unknown): Record<string, unknown>[] {
  const list = (value: unknown) => (Array.isArray(value) ? value : null);
  if (Array.isArray(payload)) return payload;
  if (!payload || typeof payload !== 'object') return [];
  const source = payload as { data?: unknown; items?: unknown };
  const items = list(source.data) ?? list(source.items)
    ?? (source.data && typeof source.data === 'object' ? list((source.data as { items?: unknown }).items) : null)
    ?? [];
  return items.filter((item): item is Record<string, unknown> => !!item && typeof item === 'object');
}

function hitShape(raw: Record<string, unknown>, position: number): SearchHit | null {
  const id = normalizeText(raw.id ?? raw.topicId ?? raw.topic_id);
  if (!id) return null;
  const rawUrl = normalizeText(raw.url ?? raw.link);
  const url = rawUrl.startsWith('https://www.cc98.org/') ? rawUrl : `https://www.cc98.org/topic/${encodeURIComponent(id)}`;
  const title = normalizeText(raw.title ?? raw.subject) || `主题 ${id}`;
  const section = normalizeText(raw.boardName ?? raw.board ?? raw.boardId);
  const publishedAt = normalizeText(raw.time ?? raw.postTime ?? raw.createTime);
  const author = normalizeText(raw.userName ?? raw.authorName ?? (raw.user as { name?: unknown } | undefined)?.name ?? raw.author);
  const replyCount = Number(raw.replyCount ?? raw.replies ?? 0) || 0;
  const candidate: Candidate = {
    sourceId: 'cc98', id, title, url,
    ...(author && { author }),
    ...(publishedAt && { publishedAt }),
    ...(section && { section }),
    ...(replyCount && { replyCount }),
  };
  return {
    candidate,
    document: { title, ...(section && { section }), ...(author && { author }), ...(publishedAt && { publishedAt }), ...(replyCount && { replyCount }) },
    position,
  };
}

export class Cc98Client {
  constructor(
    private readonly accessToken: string,
    private readonly fetch: FetchLike = (input, init) => globalThis.fetch(input, init),
  ) {}

  async searchTopics(query: string, from: number, size: number, signal?: AbortSignal): Promise<unknown> {
    const parameters = new URLSearchParams({ keyword: query, from: String(from), size: String(size) });
    let response: Response;
    try {
      response = await this.fetch(`https://api.cc98.org/topic/search?${parameters}`, {
        method: 'GET',
        credentials: 'include',
        redirect: 'follow',
        signal,
        headers: { Authorization: this.accessToken, Accept: 'application/json' },
      });
    } catch (error) {
      if (error instanceof DOMException && error.name === 'AbortError') throw error;
      throw new SourceError('无法连接 CC98，保留当前部分结果。', 'network');
    }
    if (response.status === 401) {
      throw new SourceError('请先登录 CC98，然后刷新页面再试。', 'not_logged_in');
    }
    if (response.status === 403 || response.status === 429) {
      throw new SourceError(`CC98 暂时限制了搜索请求（HTTP ${response.status}），保留当前部分结果。`, 'rate_limited');
    }
    if (!response.ok) throw new SourceError(`CC98 返回 HTTP ${response.status}，保留当前部分结果。`, 'network');
    const contentType = response.headers.get('content-type') ?? '';
    if (!contentType.includes('json')) throw new SourceError('CC98 返回内容不是 JSON，请刷新登录后重试。', 'not_logged_in');
    return response.json();
  }
}

export class Cc98SourceSession implements SearchSourceSession {
  readonly sourceId = 'cc98';
  readonly ratePolicy = CC98_RATE_POLICY;
  readonly capabilities = CC98_CAPABILITIES;
  private readonly client: Cc98Client;

  constructor(accessToken: string, fetchFn?: FetchLike) {
    this.client = new Cc98Client(accessToken, fetchFn);
  }

  async search(query: string, cursor: string | undefined, signal?: AbortSignal): Promise<SearchPage> {
    const from = cursor ? Number.parseInt(cursor, 10) : 0;
    const payload = await this.client.searchTopics(query, from, PAGE_SIZE, signal);
    const items = topicList(payload);
    const hits = items
      .map((raw, index) => hitShape(raw, from + index + 1))
      .filter((hit): hit is SearchHit => hit !== null);
    return {
      hits,
      nextCursor: items.length === PAGE_SIZE ? String(from + PAGE_SIZE) : undefined,
    };
  }
}

export const cc98Adapter: SearchSourceAdapter = {
  id: 'cc98',
  capabilities: CC98_CAPABILITIES,
  ratePolicy: CC98_RATE_POLICY,
  pageMatches: ['*://www.cc98.org/*'],
  apiHosts: ['https://api.cc98.org/'],
  createSession(pageContext: PageContext): SearchSourceSession {
    void pageContext;
    const token = readCc98AccessToken(localStorage);
    if (!token) throw new SourceError('请先登录 CC98，然后刷新页面再试。', 'not_logged_in');
    return new Cc98SourceSession(token);
  },
};
